sklearn/utils/lobpcg.py

```diff
@@ -0,0 +1,570 @@
+"""
+Locally Optimal Block Preconditioned Conjugate Gradient Method (LOBPCG)
+"""
+
+# sk-learn copy of scipy\sparse\linalg\eigen\lobpcg\lobpcg.py
```

Comment:
The current version of scipy.sparse.linalg.lobpcg is broken for small matrices and makes many of the tests in this PR fail with numeric errors. @rc and I are the authors of the scipy.sparse.linalg.lobpcg . We will take care of updating it at scipy accordingly and remove the local copy included in this PR, referring instead to scipy, as soon as the scipy version is fixed so that all tests in this PR run without errors. I cannot predict how long that would take at scipy, so have decided to include a fixed local copy of lobpcg.py into this PR so that we could finally finish the PR.

Reply:
I am aware you are the authoritative party on this algorithm and that scipy inclusion can take a while :)

The issue is that from the code quality perspective we cannot merge something that is not tested and reviewed, including the updated lobcpg implementation. Since it is destined to be in scipy, it would we somewhat un-optimal to do that in this PR (and scipy developers would be more suitable to review it).

If this PR gets merged in scikit-learn, while the lobcpg update is not yet in scipy (for any reason): possible bug regarding it, will be submitted here and from the maintenance point of view that is really something we want to avoid.

Note that the next scikit-learn release will happen in ~6 month, so that's a relatively large window for scipy inclusion. Also even assuming this gets into scipy, we would still have to backport it here, as we need support scipy 0.17.0+ in scikit-learn 0.21 (which wouldn't have it in any case).

Of course it depends what other core developers think about this and about how we can come with a reasonable solution for this.

Reply:
Of course, I understand. But if you wait for the fixed lobpcg.py to finally appear in an official release, it would probably be long enough for this PR to start conflicting with the base, because of my heavy editing of the tests sklearn\decomposition\tests\test_truncated_svd.py and sklearn\decomposition\tests\test_pca.py 

If any issues arise with lobpcg.py, I would be glad to help fixing, but keeping this PR alive for a long time, editing it every time there is a conflict with the base, does not sound like fun. 

Let me finally clarify that all the changes we have had to make in the local copy lobpcg.py in this PR fall into two categories: 

1. Fix errors handling small problems, where the actual LOBPCG algorithm is not used, but a standard LAPACK solver is called instead. Unfortunately, most of the test problems are in this category, where the scipy lobpcg fails. 

2. As we included the local copy lobpcg.py in this PR, there were many formatting and documentation errors coming from AppVeyor and Travis CI, which we fixed.

Reply:
I agree it needs testing etc. I'm fine to wait until it is merged in scipy, not released, and copy it here as long as it is tested.